naga/src/proc/constant_evaluator.rs

```diff
@@ -2497,6 +2520,121 @@ impl<'a> ConstantEvaluator<'a> {
 
         Ok(resolution)
     }
+
+    fn select(
+        &mut self,
+        reject: Handle<Expression>,
+        accept: Handle<Expression>,
+        condition: Handle<Expression>,
+        span: Span,
+    ) -> Result<Handle<Expression>, ConstantEvaluatorError> {
+        let arg = |this: &mut Self, expr| {
+            let expr = this.check_and_get(expr)?;
+            let expr = this.eval_zero_value_and_splat(expr, span)?;
+            let expr: &Expression = &this.expressions[expr];
+            this.try_eval_and_append(expr.clone(), span)
+        };
+
+        let reject = arg(self, reject)?;
+        let accept = arg(self, accept)?;
+        let condition = arg(self, condition)?;
+
+        let select_single_component =
+            |this: &mut Self, reject_scalar, reject, accept, condition| {
+                let accept = this.cast(accept, reject_scalar, span)?;
```

Comment:
I wonder, is this `cast` really necessary at all? That is, in #7572, we will end up doing all the type checking and automatic conversions in the WGSL front end, at which point this code can assume that its operands are well-typed. Then it only needs to throw errors if it encounters something it can't make progress with (like, a condition that's not `bool` or `vecN<bool>`).

Reply:
All I know is that if I _don't_ have it, the tests cases I'm using fail, even when consuming #7572. 😅

Reply:
Just confirmed that both `check_and_get` _and_ `try_eval_and_append` here weren't necessary; force-pushed f03f9bd65942.

Reply:
Concretely, this fails in constant evaluation if we don't `cast`:

```wgsl
const_assert select(42f, 9001i, true) == 9001f;
```